docs/reference/webhooks.md

```diff
@@ -2,9 +2,8 @@
 
 This section purpose is to manage the [Webhooks](https://docs.gitlab.com/ee/user/project/integrations/webhooks.html) (aka Project Hooks).
 
-The key names here are webhook URLs and values are as parameters described at [edit project hooks endpoint](https://docs.gitlab.com/ee/api/projects.html#edit-project-hook), **except the id and hook_id**.
-
-If the only value is `delete: true` then the given webhook is going to be removed.
+The key names here are webhook URLs, except if the key name is `enforce` and is set to `true` - in this case only the hooks defined here will remain in the project, all other will be deleted. Setting `enforce: false` is the same as omitting it, in which case only webhooks that have the value `delete: true` will be removed.
```

Comment:
hm... but you have one rule for all (*) of group-foo, and then you have group-bar/project-bar - would the rules from groups-foo apply to group-bar? Maybe you mean the second to be group-foo as well? But in this case I'd have thought it's obvious how it works, the "parent rule" applies until otherwise specified... 

I thought that as a user I'd like to be told explicitly that I can omit the key without unexpected consequences... but what is 'unexpected'? Maybe it's more unexpected to have hooks left over when you didn't include them in your config... OK, once again, I've written myself into it. I will omit the line :)

Reply:
I just realized I made a mistake in the example config. It was meant to be `group-foo/project-bar`.

Let me know if now that make sense.

Reply:
it does! 
but I just realised that I'm not sure what I want to happen in the above scenario: as of now, example hook gets created for project-bar, and then later deleted (as per a test I just wrote)... is that correct? It seems that all of the config beneath group-foo/* should be inherited, right? There's no point creating/updating a hook for them all, only to enforce it away again... I hadn't thought of this scenario, need to fix the inheritance.

Reply:
> It seems that all of the config beneath group-foo/* should be inherited, right?

That's correct. I think you should already get the effective inherited config in the hooks processor. Maybe check other processors to see how they handle the enforce config.

Reply:
you're right, I had made a (bad) mistake in the processor - should be fixed now, making more sense. I added a test for the inheritance scenario and some more examples in the documentation (might have gone overkill there, but it's a small section anyway, what's the harm in more info...?).

Reply:
Thanks for the extra tests and examples. Don't think there's any harm in that.

Reply:
@TigreModerata - latest update looks good. Have a question for my understanding. That's why unresolving this conversation.

What was wrong with the enforce logic at first? I don't recall seeing any issue. I see you've fixed it in the following commit by using a tupple. Could you please help me understand why/how this fixes it?

https://github.com/gitlabform/gitlabform/pull/664/commits/3561156a98ff065ff4605afbc97f4afda0884885

I don't have access to a desktop at the moment. So, it's a bit hard to try/test things out myself.

Reply:
Yeah, no problem: before it was a tuple, `config_hooks` was an iterator, `filter(..., sorted(configuration["hooks"]))` . It then got iterated over in the first loop, doing the create/update/delete. My enforce loop comes after, going through the hooks in gitlab and checking them against the config. By then, my iterator of config hooks had been consumed however, so it was checking against an empty object. My tests did not detect this, because I'm an idiot: the `enforce: true` test starts with one hook on gitlab, deletes it and creates a new one. Success :D 
After your note on inheritance, I wrote the test for it and it made it obvious.

I think I'm too eager to save whatever insignificant amounts of memory/calculation I can and tend to use iterators until they don't work.... and beyond, I guess :D

Reply:
Don't be too hard on yourself. You've been a great help. And it's awesome that you think of performance and not just what works.

> By then, my iterator of config hooks had been consumed however, so it was checking against an empty object.

Couldn't follow this. The `config_hooks` object doesn't change after going through each of the hooks and being processed. At least that's my understanding. So still not sure why the original code didn't work and how the tuple solved it.

I'm not a python expert; still learning. Appreciate walking me through this. I wonder if the schedule processor has similar bug around inheritance and enforce.

https://github.com/gitlabform/gitlabform/blob/6d012213c7144f408c74b8170f0584467c7cf628/gitlabform/processors/project/schedules_processor.py#L12

Reply:
Ah, maybe it's a python particularity: no, the sequence elements are not changed in my loop. But python iterators get "consumed" - that is, once the end of the list is reached, you'd have to reset it by hand (just redefine it, x = filter(func, list) for instance, then it starts over). The for loop is calling a "next" on the iterator, but knows to stop when no next is available. And that's when I started my new loop, over gitlab hooks, comparing them to a now empty object. E.g.:
`
>>> l = [1, 2, 3, 4, 5, 6, 7, 8]
>>> even = filter(lambda x: x%2==0, l)
>>> even
<filter object at 0x7f066a320b50>
>>> next(even)
2
>>> list(even)
[4, 6, 8]
>>> next(even)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
>>> def printeven(it):
...     for x in it:
...             print(x)
... 
>>> printeven(even)
>>> even = filter(lambda x: x%2==0, l)
>>> printeven(even)
2
4
6
8
>>> 
`

Reply:
> I'm not a python expert; still learning. Appreciate walking me through this. I wonder if the schedule processor has similar bug around inheritance and enforce.
> 
> https://github.com/gitlabform/gitlabform/blob/6d012213c7144f408c74b8170f0584467c7cf628/gitlabform/processors/project/schedules_processor.py#L12

The problem was not only with inheritance, it would have come up with other scenarios: any time I want to update a hook and apply enforce. My processor would have updated an existing gitlab hook, while consuming config_hooks. After that, it would check each existing gitlab hook against a `config_hooks`, now consumed and empty, and delete any hook not in `config_hooks = []` -> all of them.
My test worked only because I was creating a *new* hook, while deleting *all* old ones anyway. The existing gitlab hook would not have been found in `config_hooks` anyway. 
The inheritance test just made the problem explicit because of how I configured it: now there was a gitlab hook also present in config and it was getting deleted.
Lesson: test *all* cases, duh! 

I don't see anything similar in the schedules_processor, at least not an iterator problem... Does something fail there? I can take a closer look.

Reply:
Thanks for that tutorial and detailed explanation. Learned something today. Really appreciate it.

> I don't see anything similar in the schedules_processor, at least not an iterator problem... Does something fail there?

Not that I'm aware of.